doc/builders/special/runinmkshell.section.md

```diff
@@ -0,0 +1,64 @@
+# pkgs.runInMkShell {#sec-pkgs-runInMkShell}
+
+`pkgs.runInMkShell` is a function that generates a wrapper script that
+receives the command to be run as arguments and then run it inside a special
+environment created with [`mkShell`](#sec-pkgs-mkShell).
+
+It's basically like the `--run` flag of `nix-shell`, but using Nix expressions.
+
+## Usage {#sec-pkgs-runInMkShell-usage}
+
+```nix
+{ pkgs ? import <nixpkgs> { } }:
+pkgs.runInMkShell {
+  shell = pkgs.bash + "/bin/bash"
+
+  name = "demo-environment";
+
+  drv = pkgs.mkShell {
+    buildInputs = with pkgs.python3Packages; [ numpy pandas ];
+  };
+
+  prelude = ''
+    echo Hello world
+  '';
+};
+```
+
+## Attributes {#sec-pkgs-runInMkShell-attributes}
+* `shell` (default: `${pkgs.bash} + "/bin/bash"`): Which shell to use as the
+ script shebang.
+* `name` (default: `"${drv.name}-wrapper"`): Name of the derivation for the
+ script.
+* `drv`: Result of [`mkShell`](#sec-pkgs-mkShell). Can be also only the
+ attrset of the arguments for `mkShell` or the list of the packages.
+* `prelude`: Code to be pasted just before the script runs the command
+ passed with the arguments. It has access to the arguments using the
+ standard variables such as `$@`.
+
+## Function result {#sec-pkgs-runInMkShell-result}
+
+This function will always generate a executable script.
```

Comment:
It would be good to generate a script in `$out/bin`, "writeScriptBin style". That way the same function can produce for instance a file with just the environment variables for use in different shells (zsh, fish, ...), and potentially other files.
Splitting the executable and the bash library may also help with clarity.

Example files:
- `$out/bin/shell`
- `$out/shell.sh`
- `$out/shell.env`

`.env` could be done later. It's fine to limit this to bash usage for the time being.

I'd suggest not to call it `$out/bin/build-shell` etc, because that would suggest that the shell has to be closely align with some build, which it does not. These files form an interface that's more broadly applicable, including proper _development_ shells that aren't necessarily about any specific build.

Reply:
> checks if it's being sourced

Then you could symlink the `.sh` file to the executable.

My main point is that a directory can serve more purposes. For instance, `shell.env` could be the backend for `nix print-dev-env`, to support other shells, IDE extensions, ...
That's a bit more work to extract those variables, which is why I'm suggesting to do it later, but leave an "empty spot" for it in the output by making it a directory.

Reply:
`writeScriptBin` mostly accomplishes that.

I don't want Nix to impose a requirement that the script must also be source-able or anything, which is why I lean towards having separate files for these purposes.
For instance we'll probably want to do something like [`setPersonality`](https://github.com/NixOS/nix/blob/3b99c6291377cbd22607896af9dfafa857d2f2dc/src/nix/run.cc#L58) at some point, which seems like it will need a little C wrapper, which would not be natural to also support with the source-able executable pattern.

Reply:
How can I export the environment from inside the shell?

It kind of works using env but there are variables that span to multiple lines.

Reply:
`print-dev-env` dumps a bash `declare -p` after jumping through some hoops. That's hard for other software to ingest, so maybe it should be [dotenv](https://github.com/motdotla/dotenv)? Although dotenv may not actually be easy for shells to parse, paradoxically, because of `"\n"`. I really don't know what's best at this point, and I'm ok to postpone the actual `.env` file.

Reply:
I was thinking about using python to dump os.environ as json

Reply:
I was experimenting on only dumping /proc/self/environ.

It splits each variable definition with an \00 but bash is not behaving nicely when using that IFS feature in the run command.

Reply:
![image](https://github.com/NixOS/nixpkgs/assets/15693688/57a4496f-d509-4269-b5b3-cc13f93fc30a)

Reply:
got it

![image](https://github.com/NixOS/nixpkgs/assets/15693688/fe25542b-5370-4f0e-a9d6-a34fb2a23fbe)

Reply:
That still didn't get multi line values

But this works nice: https://github.com/lucasew/playground/blob/master/shell/dump-env-from-proc